streaming/base/sampling.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Functionality relating to sampling."""
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+def get_sampling(samples_per_shard: NDArray[np.int64], choose: int, granularity: int, seed: int,
+                 epoch: int, use_epoch: bool) -> NDArray[np.int64]:
+    """Get how many samples to draw from each shard of the given stream.
+
+    Args:
+        samples_per_shard (NDArray[np.int64]): Array of underlying shard sizes.
+        choose (int): How many samples to draw in total over all shards.
+        granularity (int): How many samples to draw at a time from the same shard.
+        seed (int): Seed for shuffling sampling granules.
+        epoch (int): Which epoch we are sampling for.
+        use_epoch (bool): Whether to factor epoch into the base seed, or use the same seed across
+            epochs.
+
+    Returns:
+        NDArray[np.int64]: Array of ephemeral samples chosen per shard.
+    """
+    if choose < 0:
+        raise ValueError('Choose must be a non-negative integer.')
+
+    if granularity <= 0:
+        raise ValueError('Granularity must be a positive integer.')
+
+    if seed < 0:
+        raise ValueError('Seed must be a non-negative integer.')
+
+    if epoch < 0:
+        raise ValueError('Epoch must be a non-negative integer.')
```

Comment:
Can you update the error message with "`epoch` must be a non-negeative integer, but got: {epoch}"

Reply:
done